### DISCUSSION TITLE
Table state

### DISCUSSION THREAD

**Discussion by paur94** [KevinVandy/material-react-table#469] [Q&A] [answered]
Would anyone be able to help me with saving the table in localStorage? Only if table pinning or visibility changed.

**Comment by KevinVandy** [comment, marked as answer]
There is an official example for that https://www.material-react-table.com/docs/guides/table-state-management#persistent-state

**Reply by nimo23** [reply]
@KevinVandy Would it be possible to provide a property to save the table state across re-renders:

```
    saveTableState: true|false (by default: false)
    saveTableStateLocation: sessionStorage | localStorage
```

For example:

```
// FEATURE REQUEST:
useMaterialReactTable({
    saveTableState: true
    saveTableStateLocation: sessionStorage
 }
```

Since there are dozens of settings that can be customized by the user (e.g., filtering, grouping, hiding, pagination, etc.), it would make sense to save them directly and without userland code, simply using the `saveTableState` property. You are the expert and know even the smallest settings, which should be saved in order to have a consistent/uninterrupted user experience of the table after re-renderings.

A look at the example (https://www.material-react-table.com/docs/guides/table-state-management#persistent-state) shows that such code is better placed outside the userland table, as it's a general function that can be also needed for other tables. Another advantage would be that the userland code is then freed from the table state code and can also be provided in an optimized /correct way via `saveTableState`. What do you think?

Another alternative would be to provide at least something like `table.getState().toJson()` and `table.setState()`, which can be used to (de)serialize tableState, e.g.:


```
// store the current table state to sessionStorage
sessionStorage.setItem('tableState', table.getState().toJson());

// get the current table state from sessionStorage
const currentState = JSON.parse(sessionStorage.getItem('tableState'));

// set the state from json
table.setState(currentState);
```
